ironfish/src/consensus/verifier.ts

```diff
@@ -148,29 +148,53 @@ export class Verifier {
   }
 
   /**
-   * Verify that a new transaction received over the network has valid proofs
-   * before forwarding it to the network.
-   *
-   * @params payload an unknown message payload that peerNetwork has received from the network.
-   *
-   * @returns deserialized transaction to be processed by the main handler.
+   * Verify that a new transaction received over the network can be accepted into
+   * the mempool and rebroadcasted to the network.
    */
-  verifyNewTransaction(serializedTransaction: SerializedTransaction): Transaction {
+  async verifyNewTransaction(transaction: Transaction): Promise<VerificationResult> {
     try {
-      const transaction = new Transaction(serializedTransaction)
-
       // Transaction is lazily deserialized, so we use takeReference()
       // to force deserialization errors here
       transaction.takeReference()
-
-      return transaction
+      transaction.returnReference()
     } catch (e) {
-      let message = 'Transaction cannot deserialize.'
-      if (e instanceof Error) {
-        message += ` Message: ${e.message}`
+      return { valid: false, reason: VerificationResultReason.DESERIALIZATION }
+    }
+
+    // Next, check the spends are valid and not already in the chain
+    const reason = await this.chain.db.withTransaction(null, async (tx) => {
+      const nullifierSize = await this.chain.nullifiers.size(tx)
+      const notesSize = await this.chain.notes.size(tx)
+
+      for (const spend of transaction.spends()) {
+        if (await this.chain.nullifiers.contained(spend.nullifier, nullifierSize, tx)) {
```

Comment:
I feel like we could probably reuse `verifySpend` here since it's a lot of the same logic. Maybe we can add a new verification failure for the spend.size being less than and return it in `verifySpend` then just ignore those errors. Wdyt?